modular_skyrat/modules/gpt_integration/code/gpt_subsystem.dm

```diff
@@ -0,0 +1,200 @@
+*/
+/datum/controller/subsystem/gpt/proc/converse(message, conversation_ref, role = GPT_ROLE_USER)
+	if(!api_key)
+		return
+
+	if(!conversations[conversation_ref])
+		return
+
+	if(conversation_ref in queued_responses)
+		return
+
+	var/datum/gpt_conversation/chat_holder = conversations[conversation_ref]
+
+	// Add the message to the chat holder
+	chat_holder.add_message(message, role)
+
+	queued_responses += conversation_ref
+
+	// Send the request to the GPT API
+	var/list/response = send_gpt_request(chat_holder.messages, chat_holder.endpoint_id)
+
+	queued_responses -= conversation_ref
+
+	// Add the response to the chat holder
+	for(var/response_message in response)
+		chat_holder.add_message(response_message, response[response_message])
+
+	return deformat_single_message(response)
+
+/**
+ * terminate_conversation
+ *
+ * Terminates a conversation with the GPT API and removes it from the conversation list.
+*/
+/datum/controller/subsystem/gpt/proc/terminate_conversation(conversation_ref)
+	if(!conversations[conversation_ref])
+		return
+
+	if(conversation_ref in queued_responses)
+		UNTIL(!(conversation_ref in queued_responses))
+
+	LAZYREMOVE(conversations, conversations[conversation_ref])
+
+/datum/controller/subsystem/gpt/proc/get_conversation(conversation_ref)
+	return conversations[conversation_ref]
+
+/**
+ * send_gpt_request
+ *
+ * Sends a HTTP request to the GPT API using the request header and request body procs.
+ *
+ * messages_to_send: The messages to send to the GPT API. They must be formatted correctly otherwise the request will fail.
+ * The correct format is as follows:
+ * list(list("role" = ROLE TYPE HERE, "content" = CONTENT HERE))
+ *
+ * Role is used to define where the message has come from, if it's the AI or the user.
+ *
+ * Content is the actual message that is being sent.
+ *
+ * You can have multiple messages in the list with different role types. This enables you to have advanced conversations with the AI.
+ *
+ * Naturally this is not ASYNC and will block the thread until the request is complete.
+ *
+ */
+/datum/controller/subsystem/gpt/proc/send_gpt_request(messages_to_send, endpoint_id = GPT_ENDPOINT_ID_CHAT, model_override)
+	if(!api_key)
+		return
+
+	var/datum/gpt_endpoint_config/selected_endpoint = endpoints[endpoint_id]
+
+	// Create the HTTP request
+	var/datum/http_request/request = selected_endpoint.create_http_request(messages_to_send)
+
+	// Send the request and wait for the response
+	request.begin_async()
+
+	UNTIL(request.is_complete())
+
+	return selected_endpoint.process_http_response(request.into_response())
+
+/**
+ * get_models
+ *
+ * Fetches all of the available GPT models from the OpenAI API.
+ *
+ * Returns: A list of all of the available GPT models.
+ */
+/datum/controller/subsystem/gpt/proc/get_models()
+	// Set up the required headers and URL for the GPT API
+	var/list/headers = list("Authorization" = "Bearer [api_key]")
+	var/url = "https://api.openai.com/v1/models"
```

Comment:
Why define the models route as a define if you aren't using it

Reply:
Oversight